ServiceImpl Config Class

BenWhitten-RV opened the pull request:
Implemented a config class for each ServiceImpl constructor as per the suggestion in the linked issue. I'm personally a big fan since I've been updating the constructors a fair bit recently and its a big pain to have to update every instance whenever something new is added. I'm curious what everyone else thinks though.

ctwarog-ross commented:
Alot of the instances of ServiceConfig are assigning the same values to members. What if we implemented a constructor in ServiceConfig with default values, would save some repetition

abrown-RV commented:
> Alot of the instances of ServiceConfig are assigning the same values to members. What if we implemented a constructor in ServiceConfig with default values, would save some repetition

In python they let you do something called keyword arguments where you can call any function (or ctor) with named args instead of indexed args. So you can have your defaults set in the ctor, and then override the ones you want to change.

```python
config = ServiceConfig(
    dms, # indexed since its always required
    max_connection=10,
    authz=True,
)
```
or C# lets you do
```csharp
ServiceConfig config = new ServiceConfig(dms) {
    MaxConnections = 10,
    Authz = true,
};
```

It doesn't seem like cpp supports this yet. So what I've seen is the function chaining pattern. Which is you have a bunch of setter functions for each config option that can be overridden that return a reference to itself so that you can chain the functions together.

Something like
```cpp
ServiceConfig config = ServiceConfig(dms)
    .maxConnections(10)
    .authz(true);
```

https://levelup.gitconnected.com/how-to-implement-method-chaining-in-c-3ec9f255972a

abrown-RV commented:
Sorry looking more at the PR Christian's point is about the repetitive code. My response to that would be that all the repetition is to pull the flag values in, so to take out that repetition we'd need a method that initializes a config from the flags.

My point about the function chaining is more quality of life improvement. Maybe not so necessary. I just think it looks clean when you don't have the repeated `config.` at the start of every line.

abrown-RV commented:
> C++ does have ctor, Im thinking that's how we should have this. A constructor with args and default values

Well that's no different than using the ServiceImpl ctor. The whole idea here is that the ServiceImpl ctor is growing. Even if there are nice default values, you can't take advantage of them if you want to have any non default customization.

For example with the ctor as it is on develop
```cpp
CatenaServiceImpl(std::vector<IDevice*> dms, std::string& EOPath, bool authz = false, uint16_t port = 443);
```
if you want to set the port to anything but the default, you have to set all the other args, even if you just want their defaults. And then the issue I'm raising about that is then someone has hardcoded a value to a default that may change in the future, and they don't know it.

abrown-RV commented on sdks/cpp/connections/REST/include/ServiceImpl.h:
```diff
@@ -77,6 +77,94 @@ namespace catena {
+     * @brief Sets EOPath, authz, port, and maxConnections via command line
+     * flags.
+     */
+    ServiceConfig& set_flags() {
```
Forgive my unfamiliarity with the purpose of examples. My understanding so far is that a developer working for a company actually developing a device that implements catena would create their own code that looks like our examples, but does not *use* our examples. As such I think it would be best if we did not dictate that they do flags in the same way as we do, and do not force `absl` as a dependency on them. So while I think the `set_flags` is a great idea, it would be best if specifics of our examples are separated from the sdk code. Could you do some kind of shared helper function that either builds a config from the flags and returns it, or takes an existing config and imposes the flags upon it. Looking quickly at the code base I'm not sure a good place for such a thing exists, you would know that one better.

BenWhitten-RV commented:
Removed set_flags() and just changed it back in the examples to use the setters. Clients can choose which flags to include/exclude in the business logic.
